### PR TITLE
Make `MobileBert` tokenizers independent from `Bert`

### PR DESCRIPTION
**Pull request by 501Good**
# What does this PR do?

Copied the code from `Bert` tokenizers into `MobileBert` tokenizers to make the latter self-contained.

Fixes #19303 

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

@sgugger

### CODE REVIEW

File: src/transformers/models/mobilebert/tokenization_mobilebert.py [hunk 1]
```diff
@@ -33,6 +39,471 @@
...
+def whitespace_tokenize(text):
+    """Runs basic whitespace cleaning and splitting on a piece of text."""
+    text = text.strip()
+    if not text:
+        return []
+    tokens = text.split()
+    return tokens
+
+
+# Copied from transformers.models.bert.tokenization_bert.BertTokenizer
```
Comment (sgugger) [inline src/transformers/models/mobilebert/tokenization_mobilebert.py:64]:
```suggestion
# Copied from transformers.models.bert.tokenization_bert.BertTokenizer with BERT->MobileBERT
```
The BERT in the docstring need to updated.

File: src/transformers/models/mobilebert/tokenization_mobilebert.py [hunk 2]
```diff
@@ -33,6 +39,471 @@
...
+class BertTokenizer(PreTrainedTokenizer):
```
Comment (sgugger) [inline src/transformers/models/mobilebert/tokenization_mobilebert.py:65]:
```suggestion
class MobileBertTokenizer(PreTrainedTokenizer):
```

File: src/transformers/models/mobilebert/tokenization_mobilebert.py [hunk 3]
```diff
@@ -33,6 +39,471 @@
...
+                start = end
+
+            if is_bad:
+                output_tokens.append(self.unk_token)
+            else:
+                output_tokens.extend(sub_tokens)
+        return output_tokens
+
+
 class MobileBertTokenizer(BertTokenizer):
```
Comment (sgugger) [inline src/transformers/models/mobilebert/tokenization_mobilebert.py:507]: This should be deleted now.

File: src/transformers/models/mobilebert/tokenization_mobilebert_fast.py [hunk 4]
```diff
@@ -37,6 +43,153 @@
 PRETRAINED_INIT_CONFIGURATION = {}
 
 
+# Copied from transformers.models.bert.tokenization_bert_fast.BertTokenizerFast
+class BertTokenizerFast(PreTrainedTokenizerFast):
```
Comment (sgugger) [inline src/transformers/models/mobilebert/tokenization_mobilebert_fast.py:47]:
```suggestion
class MobileBertTokenizerFast(PreTrainedTokenizerFast):
```

File: src/transformers/models/mobilebert/tokenization_mobilebert_fast.py [hunk 5]
```diff
@@ -37,6 +43,153 @@
 PRETRAINED_INIT_CONFIGURATION = {}
 
 
+# Copied from transformers.models.bert.tokenization_bert_fast.BertTokenizerFast
```
Comment (sgugger) [inline src/transformers/models/mobilebert/tokenization_mobilebert_fast.py:46]:
```suggestion
# Copied from transformers.models.bert.tokenization_bert_fast.BertTokenizerFast with BERT->MobileBERT
```

File: src/transformers/models/mobilebert/tokenization_mobilebert_fast.py [hunk 6]
```diff
@@ -37,6 +43,153 @@
...
+        if token_ids_1 is None:
+            return len(cls + token_ids_0 + sep) * [0]
+        return len(cls + token_ids_0 + sep) * [0] + len(token_ids_1 + sep) * [1]
+
+    def save_vocabulary(self, save_directory: str, filename_prefix: Optional[str] = None) -> Tuple[str]:
+        files = self._tokenizer.model.save(save_directory, name=filename_prefix)
+        return tuple(files)
+
+
 class MobileBertTokenizerFast(BertTokenizerFast):
```
Comment (sgugger) [inline src/transformers/models/mobilebert/tokenization_mobilebert_fast.py:193]: This class should be deleted.

**Review by sgugger — COMMENTED** [review]
Thanks a lot for working on this! Note that the whole goal is to adapt the copied BertTokenizer to become MobileBertTokenizer. I left comments to help achieve this :-)

File: src/transformers/models/mobilebert/tokenization_mobilebert.py (hunk 1 above)
Comment (501Good) [reply src/transformers/models/mobilebert/tokenization_mobilebert.py]: Thank you for the advice! I have updated the formatting, which should be correct now :)

**Comment by sgugger** [comment]
Hi @501Good, as you can see, your rebase has messed the diff on Git a little. Could you open a fresh PR from your branch?

**Comment by 501Good** [comment]
Hi @sgugger, sorry for that! Opened a new PR here #19531!

**Pull request closed without merging**